src/nvim/lua/executor.c

```diff
@@ -469,7 +469,10 @@ static int nlua_wait(lua_State *lstate)
   // Flush screen updates before blocking.
   ui_flush();
 
-  LOOP_PROCESS_EVENTS_UNTIL(&main_loop,
+  // Never run the main loop when we are in a fast callback.
+  Loop *loop = in_fast_callback ? NULL : &main_loop;
```

Comment:
This makes sense to me and is the general direction I hope we can drive towards. In general scripts shouldn't have to check `vim.in_fast_event()` or manually schedule() things.

Reply:
In general I think they still have to, it's just that `vim.wait()` was initially designed to be callable in fast callbacks, however there existed a path which caused it to run the main loop and crash.

Reply:
> In general I think they still have to,

yes, out of scope for this PR, but still something we can aspire to.